services/skus/service.go

```diff
@@ -637,43 +637,55 @@ func (s *Service) TransformStripeOrder(order *Order) (*Order, error) {
 
 // CancelOrder cancels an order, propagates to stripe if needed.
 //
-// TODO(pavelb): Refactor and make it precise.
-// Currently, this method does something weird for the case when the order was not found in the DB.
-// If we have an order id, but ended up without the order, that means either the id is wrong,
-// or we somehow lost data. The latter is less likely.
-// Yet we allow non-existing order ids to be searched for in Stripe, which is strange.
+// TODO(pavelb): Refactor this.
 func (s *Service) CancelOrder(orderID uuid.UUID) error {
-	// Check the order, do we have a stripe subscription?
-	ok, subID, err := s.Datastore.IsStripeSub(orderID)
-	if err != nil && !errors.Is(err, model.ErrOrderNotFound) {
-		return fmt.Errorf("failed to check stripe subscription: %w", err)
+	// TODO: Refactor this later. Now here's a quick fix.
+	ord, err := s.Datastore.GetOrder(orderID)
+	if err != nil {
+		return err
 	}
 
+	if ord == nil {
+		return model.ErrOrderNotFound
+	}
+
+	subID, ok := ord.StripeSubID()
 	if ok && subID != "" {
 		// Cancel the stripe subscription.
 		if _, err := sub.Cancel(subID, nil); err != nil {
-			return fmt.Errorf("failed to cancel stripe subscription: %w", err)
+			// Error out if it's not 404.
+			if !isErrStripeNotFound(err) {
+				return fmt.Errorf("failed to cancel stripe subscription: %w", err)
+			}
 		}
 
+		// Cancel even for 404.
 		return s.Datastore.UpdateOrder(orderID, OrderStatusCanceled)
 	}
 
-	// Try to find order in Stripe.
+	if ord.IsIOS() || ord.IsAndroid() {
+		return s.Datastore.UpdateOrder(orderID, OrderStatusCanceled)
+	}
+
+	// Try to find by order_id in Stripe.
 	params := &stripe.SubscriptionSearchParams{}
-	params.Query = *stripe.String(fmt.Sprintf("status:'active' AND metadata['orderID']:'%s'", orderID.String()))
+	params.Query = fmt.Sprintf("status:'active' AND metadata['orderID']:'%s'", orderID.String())
```

Comment:
@clD11 Decided to keep this 'active' here for now as we are not really interested in canceling already canceled subs.

What do you think?